lib/pos_handshake_controller.go

```diff
@@ -0,0 +1,173 @@
+package lib
+
+import (
+	"github.com/decred/dcrd/lru"
+	"github.com/deso-protocol/core/bls"
+	"github.com/golang/glog"
+	"math"
+)
+
+// TODO: Replace this
+func getActiveValidators() map[bls.PublicKey]*ValidatorEntry {
+	// TODO: replace with a getter to retrieve all active validators.
+	activeValidators := []*ValidatorEntry{}
+	allValidatorsMap := make(map[bls.PublicKey]*ValidatorEntry)
+	for _, validator := range activeValidators {
+		pk := validator.VotingPublicKey
+		if pk == nil {
+			continue
+		}
+		allValidatorsMap[*pk] = validator
+	}
+
+	return allValidatorsMap
+}
+
+type HandshakeController struct {
+	rnManager              *RemoteNodeManager
+	usedNonces             lru.Cache
+	protocolOnProofOfStake func() bool                              // TODO
+	getActiveValidators    func() map[bls.PublicKey]*ValidatorEntry // TODO
+}
+
+func NewHandshakeController(rnManager *RemoteNodeManager) *HandshakeController {
+
+	vm := &HandshakeController{
+		rnManager:              rnManager,
+		usedNonces:             lru.NewCache(1000),
+		protocolOnProofOfStake: func() bool { return false },
+	}
+
+	return vm
+}
+
+func (hc *HandshakeController) InitiateHandshake(rn *RemoteNode) {
+
+	nonce := uint64(RandInt64(math.MaxInt64))
+	if err := rn.InitiateHandshake(nonce); err != nil {
+		glog.Errorf("RemoteNode.InitiateHandshake: Error initiating handshake: %v", err)
+		hc.rnManager.Disconnect(rn)
+	}
+}
+
+func (hc *HandshakeController) handlePoSHandshakePeerMessage(remoteNode *RemoteNode) {
+	if !hc.protocolOnProofOfStake() {
+		return
+	}
+
+	// Make sure the peer is on the right proof of stake version
+	if remoteNode.GetNegotiatedProtocolVersion() != ProtocolVersion2 {
+		// Disconnect the peer because we only accept validators running proof of stake.
+		hc.rnManager.Disconnect(remoteNode)
+		return
+	}
+
+	// Get all active validators and see if the peer is one of them.
+	activeValidators := hc.getActiveValidators()
+	validatorPk := remoteNode.GetValidatorPublicKey()
+	if validatorPk == nil {
+		// Disconnect the peer because we only accept validators running proof of stake.
+		hc.rnManager.Disconnect(remoteNode)
+		return
+	}
+
+	// If there's already a validator connected with the same public key, disconnect the peer.
+	if _, ok := hc.rnManager.GetValidatorIndex().Get(validatorPk.Serialize()); ok {
+		hc.rnManager.Disconnect(remoteNode)
+		return
+	}
+
+	// If the peer is not an active validator, there is nothing else to check so return.
+	if _, ok := activeValidators[*validatorPk]; !ok {
+		return
+	}
+}
+
+func (hc *HandshakeController) _handleHandshakePeerMessage(origin *Peer, desoMsg DeSoMessage) {
```

Comment:
When is this handler called? Not 100% based on the naming. Looks like it's called when the handshake has completed. It will be good rename, if so